lib/compat/wordpress-6.1/rest-api.php

```diff
@@ -93,3 +93,84 @@ function gutenberg_register_has_archive_on_post_types_endpoint() {
 	);
 }
 add_action( 'rest_api_init', 'gutenberg_register_has_archive_on_post_types_endpoint' );
+
+/**
+ * Use custom REST API Controller for Navigation Posts
+ *
+ * @param array $args the post type arguments.
+ * @return array the filtered post type arguments with the new REST Controller set.
+ */
+function gutenberg_update_navigation_rest_controller( $args, $post_type ) {
+	if ( in_array( $post_type, array( 'wp_navigation' ), true ) ) {
+		// Original set in
+		// https://github.com/WordPress/wordpress-develop/blob/6cbed78c94b9d8c6a9b4c8b472b88ee0cd56528c/src/wp-includes/post.php#L528.
+		$args['rest_controller_class'] = 'Gutenberg_REST_Navigation_Controller';
+	}
+	return $args;
+}
+add_filter( 'register_post_type_args', 'gutenberg_update_navigation_rest_controller', 10, 2 );
+
+function gutenberg_transform_slug_to_post_id( $response, $handler, WP_REST_Request $request ) {
+
+	// Create (POST) requests send the `slug` as part of the request **body** for the Post to be created.
+	// As a result the `slug` in POST requests should not be transformed into an ID.
+	$should_preserve_slug = 'POST' === $request->get_method();
+
+	// Ignore non-Navigation REST API requests.
+	if( ! str_starts_with( $request->get_route(), '/wp/v2/navigation' ) || $should_preserve_slug )  {
+		return $response;
+	}
+
+	// Get the slug from the request.
+	$slug = $request['slug'];
+
+	// If no slug provided assume ID and continue as normal.
+	if( empty($slug) ) {
+		return $response;
+	}
+
+	$args = array(
+		'name'                   => $slug, // query by slug
+		'post_type'              => 'wp_navigation',
+		'nopaging'               => true,
+		'posts_per_page'         => '1',
+		'update_post_term_cache' => false,
+		'no_found_rows'          => true,
+	);
+
+	// Query for the Navigation Post by slug (post_name).
+	$query = new WP_Query( $args );
+
+	if ( empty( $query->post->ID ) ) {
+		return new WP_Error(
+			'rest_post_not_found',
+			__( 'No navigation found.' ),
+			array( 'status' => 404 )
+		);
+	}
+
+	// Set the post ID based on the slug.
+	$request['id'] = $query->post->ID;
+
+	return $response;
+}
+add_filter('rest_request_before_callbacks', 'gutenberg_transform_slug_to_post_id', 10, 3 );
+
+
+
+
+function gutenberg_update_navigation_rest_route_for_post($route, WP_Post $post) {
+
+	if( $post->post_type !== 'wp_navigation' ) {
+		return $route;
+	}
+
```

Comment:
```suggestion
	$post_type_route = rest_get_route_for_post_type_items( $post->post_type );
	if ( ! $post_type_route ) {
		return '';
	}

	return sprintf( '%s/%s', $post_type_route, $post->slug );
```

I think this is preferrable over a `preg_replace` which won't account for changings to the collection base.

Reply:
This is much better. Thanks. 

It did introduce a minor bug due to the usage of `$post->slug` but the test caught that and I've fixed it.

Reply:
I can't be trusted to write code without autocomplete 😀